lib/writable.js

```diff
@@ -1,69 +1,227 @@
+    this.writableLength = 0;
+    return ret;
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    defaultEncoding = 'utf-8',
+    highWaterMark = DEFAULT_HIGH_WATERMARK
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.highWaterMark = highWaterMark;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(highWaterMark);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked) this.corked--;
+    if (!this.corked && this.writableLength) {
+      this.writing = true;
+      this.doWrite(this.buffer.getBuffer());
+    }
+  }
+
+  setDefaultEncoding(encoding) {
+    this.defaultEncoding = encoding;
+  }
+
+  error(err, cb) {
+    if (cb && typeof cb === 'function') {
+      process.nextTick(cb, err);
+    }
+    this.emit('error', err);
+  }
+
+  pipe() {
+    const err = new Error('[ERR_STREAM_CANNOT_PIPE]');
+    this.error(err);
+  }
+
+  validChunk(chunk, cb) {
+    let error;
+    if (chunk === null) {
+      error = new Error('[ERR_STREAM_NULL_VALUES]');
+    } else if (typeof chunk !== 'string' && !Buffer.isBuffer(chunk)) {
+      error = new Error('[ERR_INVALID_ARG_TYPE]');
+    }
+
+    if (error) {
+      this.error(error, cb);
+      return false;
+    }
+    return true;
+  }
+
+  write(chunk, encoding, cb) {
+    if (typeof encoding === 'function') {
+      cb = encoding;
+      encoding = null;
+    }
+
+    if (this.ended) {
+      const err = new Error('[ERR_STREAM_WRITE_AFTER_END]');
+      this.error(err, cb);
+      return;
+    }
+
+    if (this.destroyed) {
+      const err = new Error('[ERR_STREAM_DESTROYED]');
+      this.emit(err, cb);
+      return;
+    }
+
+    if (!encoding) {
+      encoding = this.defaultEncoding;
+    }
+
+    if (!this.validChunk(chunk, cb)) return;
+    if (cb) this.callbacks.push(cb);
+
+    if (this.corked || this.writing) {
+      this.buffer.add(chunk, encoding);
+    } else {
+      if (typeof chunk === 'string') {
+        chunk = Buffer.from(chunk, encoding);
+      }
+      this.writing = true;
+      this.doWrite(chunk);
+    }
+  }
+
+  doWrite(data) {
+    const callbacks = this.callbacks;
+    this.callbacks = [];
+
+    const cb = () => {
+      callbacks.forEach(cb => cb());
+      if (!this.writableLength) {
+        this.writing = false;
+        if (this.ending) {
+          this.ended = true;
+          this.emit('finish');
+        }
+        return;
+      }
+      this.doWrite(this.buffer.getBuffer());
+    };
+
+    if (Array.isArray(data)) {
+      if (this._writev) this._writev(data, cb);
+      else this._write(Buffer.concat(data), cb);
+    } else {
+      this._write(data, cb);
+    }
```

Comment:
Based on this usage I think both _write and _writev are required then? Otherwise, you should accommodate for one of them being missing.